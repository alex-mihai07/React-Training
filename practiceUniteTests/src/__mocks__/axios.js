import {countriesMock} from "../utils/tests/global-mocks";

const axios = {
    get: jest.fn().mockResolvedValue({
        data: [
            ...countriesMock
        ]
    })
}

export default axios