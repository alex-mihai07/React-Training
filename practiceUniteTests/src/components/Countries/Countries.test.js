import React from 'react'
import {render, fireEvent, act, waitFor} from '@testing-library/react'
import'@testing-library/jest-dom'
import Countries from "./Countries"
import axios from 'axios'


const mockURL = 'https://restcountries.eu/rest/v2';
const apiRejectPromise = jest.fn(() => Promise.reject(() => new Error('my own error')))

describe("Countries component", () => {
    const mockOnChange = jest.fn()


    it("should match snapshot", () => {
        const {asFragment} = render(<Countries/>)

        expect(asFragment()).toMatchSnapshot();
    })

    it("should handle onChange query functionality", () => {
        const {asFragment, getByTestId} = render(<Countries/>)
        const mockedNewValue = "Slovakia"
        act(() => {

            fireEvent.change(getByTestId("countries-search-query"), {

                target: {value: mockedNewValue},

            });
        })

        const newInputValue = getByTestId("countries-search-query").value
        expect(newInputValue).toBe(mockedNewValue)
    })

    //same goes for select change query

    it('should call the data fetch API', async () => {
        //render the component
        const { getByTestId } = render(<Countries/>)
        const mockedNewValue = "Romania"

        //mock the fetch
        //act fetch
        await act(async () => {
             await fireEvent.change(getByTestId("countries-search-query"), {
                target: {value: mockedNewValue},
            });
        })

        const queryValue = getByTestId('countries-search-query').value
        const finalUrl = `${mockURL}/name/${queryValue}?fields=name;capital;region;population;flag`

        await waitFor(() => expect(axios.get).toHaveBeenCalledWith(finalUrl))


        //expect state to be populated
    })

    it('should display the loader when fetching data', async () => {
        //render the component
        const { getByTestId } = render(<Countries/>)
        const mockedNewValue = "Romania"

        //mock the fetch
        //act fetch
        act(() => {
            fireEvent.change(getByTestId("countries-search-query"), {
                target: {value: mockedNewValue},
            });
        })

        await waitFor(() => expect(getByTestId('loader')).toBeInTheDocument())
    })

    it('should display error when fetching data from API fails', async () => {
        //render the component
        const { getByTestId } = render(<Countries/>)
        const mockedNewValue = "Romania"

        //option 1
        // axios.get.mockImplementationOnce(apiRejectPromise)

        //option 2
         axios.get.mockRejectedValueOnce({message: 'custom error'})

        //mock the fetch
        //act fetch
         act( () => {
             fireEvent.change(getByTestId("countries-search-query"), {
                target: {value: mockedNewValue},
            });
        })

        expect(getByTestId('error')).toBeInTheDocument()
        // await waitFor(() => expect(getByTestId('error')).toBeInTheDocument())
    })

    it('should display the countries correctly', async () => {
        //render the component
        const { getByTestId } = render(<Countries/>)
        const mockedNewValue = "Romania"

        //option 1
        // axios.get.mockImplementationOnce(apiRejectPromise)

        //option 2
        axios.get.mockRejectedValueOnce({message: 'custom error'})

        //mock the fetch
        //act fetch
        act( () => {
            fireEvent.change(getByTestId("countries-search-query"), {
                target: {value: mockedNewValue},
            });
        })

        expect(getByTestId('error')).toBeInTheDocument()
        // await waitFor(() => expect(getByTestId('error')).toBeInTheDocument())
    })
})