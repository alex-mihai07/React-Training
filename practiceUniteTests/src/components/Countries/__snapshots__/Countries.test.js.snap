// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countries component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <div
      class="filters"
    >
      <input
        data-testid="countries-search-query"
        placeholder="Search for a country..."
        type="text"
        value=""
      />
      <select
        class="selectFilter"
        name="region"
        placeholder="Filter by region"
      >
        <option
          value=""
        >
          Filter by region
        </option>
        <option
          value="africa"
        >
          Africa
        </option>
        <option
          value="americas"
        >
          America
        </option>
        <option
          value="asia"
        >
          Asia
        </option>
        <option
          value="europe"
        >
          Europe
        </option>
        <option
          value="oceania"
        >
          Oceania
        </option>
      </select>
    </div>
    <div
      class="results"
    />
  </div>
</DocumentFragment>
`;
